3.13/redundancy_eliminator.md

```diff
@@ -0,0 +1,86 @@
+# The Tier 2 Redundancy Eliminator
+
+Author: Ken Jin, Mark Shannon
+
+Tier 2 traces require optimization. At minimum, redundant operations
+should be removed. This is where the name "Redundancy Eliminator" is
+derived from. This forms part of the larger effort towards
+the tier 2 optimizer.
+
+## Abstract Interpretation of Uops
+
+To analyze the uops, [abstract interpretation](https://en.wikipedia.org/wiki/Abstract_interpretation)
+shall be used.
+
+Just like the CPython interpreter is an interpretation over
+Python program state, represented by `PyObject *`, the abstract interpreter
+shall be an interpretation over Python symbolic information state,
+represented by `_Py_UOpsSymType *`.
+
+### Lattice
+
+The states of information in such an abstract interpretation can be
+represented by a [lattice](https://en.wikipedia.org/wiki/Lattice_(order))
+of values. The lattice itself is an ordered powerset
+of all variables and their associated information.
+
+For example, here is a trivial lattice for the statement 
+```python
+x = 1
+y = 2.0
+```
+:
+
+```mermaid
+flowchart BT
+    A("`Bottom, ie. x: *?*, y: *?*`") --> B("`x: *int*, y: *?*`")
+    A --> C("`x: *?*, y: *float*`")
+    B --> D("`Top, x: *int*, y: *float*`")
+    C --> D
```

Comment:
If it wasn't apparently the standard way to render lattices (at least, all the diagrams in the Wikipedia article also use it), it would be more understandable if the chart was rendered upside down, so the flow corresponds to the code... Not sure what should prevail, conformance to the theory or to the code.

Reply:
Lattices in compilers are usually rendered bottom-up, so a little different from the math version in wikipedia.

Reply:
Oh apparently it's normally top down. Let me reformat.